File: src/main/java/org/telran/lecture_14_hash/examples/HashExamples.java
package org.telran.lecture_14_hash.examples;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

public class HashExamples {

    public static int hashKeyMode(String key, int tableSize) {
        int hash = 0;
        for (int i = 0; i < key.length(); i++) {
            hash += key.charAt(i);
        }
        return hash % tableSize;
    }

    public static int multiplicationHash(String key, int tableSize) {
        int hash = 0;
        for (int i = 0; i < key.length(); i++) {
            hash += key.charAt(i);
        }
        double A = 0.6180339887; // Константа (золотое сечение)
        double fractionalPart = (hash * A) % 1;
        return (int) Math.floor(fractionalPart * tableSize);
    }

    public static String sha256Hash(String data) throws NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] encodedHash = digest.digest(data.getBytes(StandardCharsets.UTF_8));
        return bytesToHex(encodedHash);
    }

    public static String sha1Hash(String data) throws NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance("SHA-1");
        byte[] encodedHash = digest.digest(data.getBytes(StandardCharsets.UTF_8));
        return bytesToHex(encodedHash);
    }

    private static String bytesToHex(byte[] hash) {
        StringBuilder hexString = new StringBuilder(2 * hash.length);
        for (byte b : hash) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }
        return hexString.toString();
    }

    public static void main(String[] args) throws NoSuchAlgorithmException {
        String myData = "Hello, world!";
        String hashedData1 = sha1Hash(myData);
        String hashedData256 = sha256Hash(myData);
        System.out.println("hashedData1 =      " + hashedData1);
        System.out.println("hashedData256 =    " + hashedData256);

        String[] keys = {"Hello", "Hell", "hell", "help", "Hi", "Alex", "Ivan", "John", "Kapibara", "My name is", "Abu", "key", "22"};
        int tableSize = 100;

        for (String key : keys) {
            System.out.println("hash_key_mode(\"" + key + "\", " + tableSize + ") = " + hashKeyMode(key, tableSize));
        }
        System.out.println("-------------------------------------");
        for (String key : keys) {
            System.out.println("multiplicationHash(\"" + key + "\", " + tableSize + ") = " + multiplicationHash(key, tableSize));
        }
    }
}

File: src/main/java/org/telran/lecture_14_hash/examples/HashExamples.js
const crypto = require('crypto');


function hash_key_mode(key, tableSize) {
    if (typeof key !== 'number') {
        // Преобразуем строку в число (простой пример)
        let hash = 0;
        for (let i = 0; i < key.length; i++) {
            hash += key.charCodeAt(i);
        }
        key = hash;
    }
    return key % tableSize;
}

function multiplicationHash(key, tableSize) {
    if (typeof key !== 'number') {
        // Преобразуем строку в число (простой пример)
        let hash = 0;
        for (let i = 0; i < key.length; i++) {
            hash += key.charCodeAt(i);
        }
        key = hash;
    }
    const A = 0.6180339887; // Константа (золотое сечение)
    const fractionalPart = (key * A) % 1;
    return Math.floor(fractionalPart * tableSize);
}

function sha256Hash(data) {
    const hash = crypto.createHash('sha256');
    hash.update(data);
    return hash.digest('hex');
}

function sha1Hash(data) {
    const hash = crypto.createHash('sha1');
    hash.update(data);
    return hash.digest('hex');
}


const myData = 'Hello, world!';
const hashedData1 = sha1Hash(myData);
const hashedData256 = sha256Hash(myData);
console.log(`hashedData1 =      ${hashedData1}`);
console.log(`hashedData256 =    ${hashedData256}`);


// const simpleHashedData2 = sha256Hash(myData);
const keys = ["Hello", "Hell", "hell", "help", "Hi", "Alex", "Ivan", "John", "Kapibara", "My name is", "Abu", "key", "22"];
const tableSize = 100;

for (const key of keys) {
    console.log(`hash_key_mode("${key}", ${tableSize}) = ${hash_key_mode(key, tableSize)}`);
}
console.log("-------------------------------------")
for (const key of keys) {
    console.log(`multiplicationHash("${key}", ${tableSize}) = ${multiplicationHash(key, tableSize)}`);
}

let myObject = {};
myObject["key1"] = "value1";
myObject[42] = "value2";
console.log(myObject["key1"])

let myMap = new Map();
let objKey = { id: 1 };
myMap.set("key1", "value1");
myMap.set(objKey, "value2");
console.log(myMap.get(objKey));